### ISSUE TITLE
Player accessibility improvements.

### ISSUE THREAD

**Issue by FrenjaminBanklin** [ucfopen/enigma-materia-widget#41]
- ~~An element exists in the DOM which contains keyboard use instructions, but it is not reachable in the tab order.~~
  - ~~It's easy enough to add this element to the tab order, but for users who have to use a keyboard but do not use a screen reader it ends up being a mysterious invisible extra tab stop.~~
    - Replace this always-invisible element with something that is only visible when it has focus and shows/hides a dialog window with the keyboard instructions. Sighted keyboard users can activate the element like a button to bring up a dialog window with keyboard instructions, and we can aria-label the element with the full instructions for unsighted users so they can tab past it without needing to open the dialog.

- ~~It's possible to tab out of an active question because the header and game board are still rendered and still in the tab order.~~
  - ~~Either make everything outside of the question popup unreachable by tab, or opt not to render them at all while there's a current question.~~
    - ~~This might be problematic since it looks like the questions still being tab-available is what's allowing the focus to be returned to the game board after a question popup is dismissed.~~
  - This can be handled by setting a value for the `inert` attribute on the header and game board elements when a question is selected, then removing that attribute from those elements when a question is cleared. This requires that the `id="checkAns"` element being read by screen readers immediately upon answering a question must be moved out of the header to remain visible while the header is inert.

- ~~The question text is not reachable in the tab order, so it's only read once to screen readers and can not be repeated.~~
  - This is resolved by adding the question text to the tab order, and also underlining the text when it has focus.

- ~~Answer options within a question have two tab-reachable elements - the `li` itself and the nested `input`.~~
  - ~~Potential solution may be to remove the radio button from the tab order and make sure the `li` is labeled properly for interpretation by screen readers. If removing the radio button from the tab order, use aria roles so the `li`s are treated like radio buttons, and make sure up/down arrow key presses move between options appropriately. The ideal solution is probably to remove the `li` from the tab order and allow the radio buttons to function as the screen reader expects them to.~~
    - This is resolved by removing the nested inputs from the tab order and using a combination of new up/down arrow key listeners and aria roles to recreate the natural behavior of radio groups.
    - Would also be nice if pressing Enter moved focus to the button to submit, rather than forcing users to tab through any remaining answer options and then past the 'Return' button to reach it.
      - We could potentially move focus to the submission button if Enter or Space are pressed on an answer that is selected already, but that may be too great a departure from the usual behavior of radio groups.

- ~~Leaving a question without answering it requires tabbing through all of the possible answers to reach the 'Return' button.~~
  - This is improved by adding a listener for the Escape key that allows users to leave a question immediately.

- ~~Submitting an answer will indicate how many unanswered questions remain and will indicate whether the answer just submitted was correct or incorrect, but there is no way of reaching an element on the page at will which would indicate the current score or how many questions remain unanswered.~~
  - Explicitly setting the questions remaining/current score element's tab index to 0 allows it to be reached naturally. ~~This seems to incur a new (possibly non-issue?) issue: since the score element is within the header, the header's text content is always read out after the score element's content is read out.~~
    - This is resolved by making both the h1 heading element containing the text and the questions remaining element both individually reachable in the tab order, rather than the whole header element itself. This allows the question text to be re-read, and also allows the questions remaining text to be read without also re-reading the widget title every time.

- Pressing the Enter/Return key when the 'Return' button inside a question is focused leaves the current question (expected) and sometimes selects whichever question has focus in the main area (unexpected). Pressing the Space key does not do this.
  - ~~Make sure there aren't redundant KeyUp or KeyPress listeners and that keypress events aren't bubbling further than they have to.~~
  - There might not be anything we can do about this, since the 'click' handler is also activated by the Space and Enter keys. In the case of the Enter key, keeping the key held down seems to fire multiple keypress events whereas with the Space key, there's only one. The issue actually appears to be one 'Enter' keypress triggering the event on the 'Return' button and then, if the key is held down, triggering a second event on the question element in the game board after it receives focus. I'm not sure if it's possible (or worthwhile if so) to revisit the event handling on the game board question elements to only react to a keyup event where there was a corresponding keydown event.
    - Maybe using a keydown event on the question element to 'prime' it for a keyup event? It feels like a bit of an overengineered solution for what is ultimately not a significant problem.

- ~~When submitting an answer, the 'Submit Final Answer' button retains focus but is no longer interactive. The 'Submit Final Answer' button is also no longer reachable with tab after an answer has been submitted.~~
  - ~~Either auto-focus the 'Return' button after submitting an answer, or change the behavior of the 'Submit Final Answer' button after submitting an answer to function like the 'Return' button does. No longer being able to reach it via tab is probably fine in either case.~~
  - Auto-focusing the 'Return' button and modifying the text read by screen readers to 'Press the Space or Enter key to...' rather than 'Tab back to the Return button to...' seems like a smoother experience.

- ~~Audio elements used as media within a question require an additional tab to reach the interactive controls.~~
  - ~~Adjust tab order so that the audio element's parent isn't reachable via tab but the audio element itself is?~~
  - It looks like the `audio` DOM element is receiving focus first, and then subsequent tabs move through the interactive controls so there may not be anything we can do about this behavior except add some kind of visual indicator when the audio element (or any of its controls) have focus, so sighted keyboard users know what's going on.
    - Of course, the audio controls look different for every browser so there's only so much we can do to try standardizing the appearance when they have focus.

- ~~Images used as media within a question require an additional tab to reach the 'expand' button. Once an image has been expanded the 'collapse' button is not reachable in the tab order.~~ Images used as media allow for optional image descriptions, which are rendered separately in an adjacent span but are not used as alt text.
  - ~~Adjust tab order so that the expand button's parent isn't reachable via tab but the button itself is? For the purpose of screen readers, the image itself should probably be reachable as well. The 'close' button should at the very least be in the tab order, but it should probably also receive focus automatically after the expand button is activated.~~
  - Removing the parent div from the tab order and adding the image itself to the tab order makes a bit more sense, also adding a visual effect to indicate when the image is focused to help sighted keyboard users.
    - A combination of setting 'inert' on the non-lightbox parts of the question popup when media is expanded, auto-focusing the image/video/whatever with a directive, adjusting tab indices and adding visual indicators for focus should hopefully resolve any other lingering keyboard navigation difficulties. Also auto-focusing the question text after closing the lightbox.
  - The missing alt text may be tricky, since it's kind of on the widget authors to provide it and it could potentially be different from the 'description' text which is already optional. We could potentially add a second input for alt text in the creator and mark the question as incomplete/erroneous if alt text isn't provided.

- ~~Numerous things visible on the page can be reached in the tab order but have no visual indicator that they are currently the element of focus.~~
  - This is easy enough to solve by adding a glow effect using a centered box shadow to the score wheel and category heading elements when they receive focus. The same effect can be added to the widget title using a text shadow.

- It would be nice to be able to move around the game board more freely. Currently the only way to navigate the game board with a keyboard is to tab through all of the questions left to right top to bottom in order, ~~and if any questions were skipped returning to the top requires shift-tabbing backwards through every question and category heading~~.
  - Add listeners to the game board itself for the arrow keys, allow arrow keys to move focus to the appropriate question.
    - It may be necessary to read out instructions per element as to what the arrow keys will do, but that may also be overkill?
  - A button can be rendered after the list of categories which has no width or height and is not visible until it is given focus, meaning it can only be interacted with specifically by tabbing to it. This button's space/enter keypress/click behavior can move focus back to the top of the game board as if it had been tabbed into from the score indicator normally. It may make sense to place an equivalent button at the 'top' of the game board, before any categories are rendered, that moves focus to bottom of the game board - either all the way to the bottom-rightmost question or only to the bottommost category heading.
